File: 15 3Sum.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        leng = len(nums)
        re = []
        if leng > 2:
            for i in range(leng-2):
                if nums[i] > 0:
                    break;
                if i != 0 and nums[i] == nums[i-1]:
                    continue;
                l = i + 1
                r = leng - 1
                while l < r:
                    if nums[i] + nums[l] + nums[r] == 0:
                        re.append([nums[i], nums[l], nums[r]])
                        while l < r and nums[l] == nums[l+1]:
                            l += 1
                        while l < r and nums[r] == nums[r-1]:
                            r -= 1
                        l += 1
                        r -= 1
                    elif nums[i] + nums[l] + nums[r] > 0:
                        r -= 1;
                    else:
                        l += 1
        return re

File: 15 3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int leng = nums.size();
        vector<vector<int>> re;
        if(leng > 2){
            for(int i = 0;nums[i] <= 0 && i < leng-2;++i){
                if(i != 0 && nums[i] == nums[i-1])
                    continue;
                int l = i+1, r = leng-1;
                while(l < r){
                    if(nums[i] + nums[l] + nums[r] == 0){
                        re.push_back({nums[i], nums[l], nums[r]});
                        while(l < r && nums[l] == nums[l+1])    ++l;
                        while(l < r && nums[r] == nums[l-1])    --r;
                        ++l, --r;
                    }
                    else if(nums[i] + nums[l] + nums[r] > 0){
                        --r;
                    }
                    else{
                        ++l;
                    }
                }
            }
        }
        return re;
    }
};